### PR TITLE
S3 support

### LINKED ISSUE #96: Support for ESP32-S3
since it is already supported in Arduino ESP32 git branch https://github.com/espressif/arduino-esp32/tree/esp32-s3-support
and a official release for will be soon done.

### PR DESCRIPTION
**Pull request by Jason2866**
fixes #96 

@balloob can you please build a new release and provide a NPM package? So we can use with Webtools.
THX!

### CODE REVIEW

File: src/const.ts [hunk 1]
```diff
@@ -157,7 +190,7 @@ export const CHIP_ERASE_TIMEOUT = 600000; // timeout for full chip erase in ms
 export const MAX_TIMEOUT = CHIP_ERASE_TIMEOUT * 2; // longest any command can run in ms
 export const SYNC_TIMEOUT = 100; // timeout for syncing with bootloader in ms
 export const ERASE_REGION_TIMEOUT_PER_MB = 30000; // timeout (per megabyte) for erasing a region in ms
-export const MEM_END_ROM_TIMEOUT = 50;
+export const MEM_END_ROM_TIMEOUT = 500;
```
Comment (balloob) [inline src/const.ts:193]: 10x timeout?

File: src/esp_loader.ts [hunk 2]
```diff
@@ -17,7 +19,7 @@ import {
   ESP_READ_REG,
   ESP_WRITE_REG,
   ESP_SPI_ATTACH,
-  ESP_SPI_SET_PARAMS,
+  //ESP_SPI_SET_PARAMS,
```
Comment (balloob) [inline src/esp_loader.ts:22]:
```suggestion
```

File: src/esp_loader.ts [hunk 3]
```diff
@@ -57,7 +58,7 @@ export class ESPLoader extends EventTarget {
   _efuses = new Array(4).fill(0);
   _flashsize = 4 * 1024 * 1024;
   debug = false;
-  IS_STUB = false;
+  IS_STUB = true;
```
Comment (balloob) [inline src/esp_loader.ts:61]: The first time it loads the stub is not loaded. Why set this to true?

File: src/esp_loader.ts [hunk 4]
```diff
@@ -670,15 +669,15 @@ export class ESPLoader extends EventTarget {
     let flashWriteSize = this.getFlashWriteSize();
     if (
       !this.IS_STUB &&
-      [CHIP_FAMILY_ESP32, CHIP_FAMILY_ESP32S2].includes(this.chipFamily)
+      [CHIP_FAMILY_ESP32, CHIP_FAMILY_ESP32S2, CHIP_FAMILY_ESP32S3, CHIP_FAMILY_ESP32C3].includes(this.chipFamily)
     ) {
       await this.checkCommand(ESP_SPI_ATTACH, new Array(8).fill(0));
     }
-    if (this.chipFamily == CHIP_FAMILY_ESP32) {
+    //if (this.chipFamily == CHIP_FAMILY_ESP32) {
```
Comment (balloob) [inline src/esp_loader.ts:676]: Remove this code if it's no longer needed.

File: src/esp_loader.ts [hunk 5]
```diff
@@ -768,7 +767,9 @@ export class ESPLoader extends EventTarget {
   getBootloaderOffset() {
```
Comment (balloob) [inline src/esp_loader.ts:767]: We have so many chip family checks in supported methods like mac address or bootloader offset, that it feels like we should consider adding this extra info and merge it with output from `getSpiFlashAddresses`. (not in this PR though)

File: src/esp_loader.ts [hunk 6]
```diff
@@ -917,12 +918,16 @@ export class ESPLoader extends EventTarget {
       let words = unpack("I".repeat(Math.floor(data.length / 4)), data);
       let nextReg = SPI_W0_REG;
 
-      this.logger.debug(`Words Length: ${words.length}`);
+      if (this.debug) {
+        this.logger.debug(`Words Length: ${words.length}`);
```
Comment (balloob) [inline src/esp_loader.ts:922]: It's interesting we have both `this.debug` and I bet that `this.logger.debug` has it's own filtering … 🤷

File: src/const.ts (hunk 1 above)
Comment (Jason2866) [reply src/const.ts:193]: Had problems. Used the value from https://github.com/adafruit/Adafruit_WebSerial_ESPTool/blob/main/js/esptool.js#L107

File: src/esp_loader.ts (hunk 3 above)
Comment (Jason2866) [reply src/esp_loader.ts]: true (left over from tests)

File: src/esp_loader.ts (hunk 4 above)
Comment (Jason2866) [reply src/esp_loader.ts]: done

**Review by balloob — APPROVED** [review]

**Pull request merged**